README.md
# 元数据总结

- [元数据总结](#元数据总结)
  - [关于 Big Endian 和 Little Endian](#关于-big-endian-和-little-endian)
  - [JPEG](#jpeg)
  - [PNG](#png)
  - [Webp](#webp)
  - [Gif](#gif)

## 关于 Big Endian 和 Little Endian

写在前面

计算机中储存数值有 2 种方式，一种是大端存储（Big Endian），一种是小端存储（Little Endian）。

以 uint32 数值的储存为例来举例：

所谓大端储存就是指存储的数值是从高位到低位的顺序，，就是数值占用 4 个字节，比如：

0x00 0x00 0x00 0x01 转为十进制，结果是 1

小端储存就不同了，同样的数，如果是小端储存，高位存储在低地址，低位存储在高地址，比如：

0x00 0x00 0x00 0x01 转为十进制，结果是 256^3。等于大端储存的 0x01 0x00 0x00 0x00

## JPEG

因为历史原因，也叫 JPG

储存方式有如下的规则，其中一般使用开头和结尾的各 2 个字节来判断是否为 JPEG 格式

开头: 0xFF 0xD8
结尾: 0xFF 0xD9

如下所示：

```tree
| start: 2 bytes | block | block | ... | end: 2 bytes |
```

然后，储存时是按数据块（block）来储存的，每个数据库块的开头都有一个 2 字节的标识符，为 0xFF，type，来标识块的类型。
其中 C0 类型为图片的宽高信息。

数据块分析:

| 数值 | 偏移量      | 长度 | 说明                                       |
| ---- | ----------- | ---- | ------------------------------------------ |
| FF   | +0          | 1    | 开始标识符                                 |
| 任意 | +1          | 1    | 类型，查看 JPEG 规范                       |
| n    | +2 ~ +3     | 2    | 块长度，包含这两位，但不包含 FF 和类型长度 |
| 内容 | +4 ~ +(n+2) | n-2  | 内容                                       |

---

已知类型标识符

| 数值 | 说明     | 信息                      |
| ---- | -------- | ------------------------- |
| C0   | 宽度高度 | `+5~+6：高度 +7~+8：宽度` |

## PNG

开头固定为:

`89 50 4e 47 0d 0a 1a 0a 00 00 00 0d 49 48 44 52`

宽度在 0x00000010 偏移量处, `0~3` 位是宽度 `4~7` 位是高度

但是按照 PNG 规范，其实表示宽度和高度的字符各占 4 个字节，但实际应用中 2 字节可以表示的最大数值为 256^2-1=65535，而当前实际应用中我没有发现有这么大的图片，所以这里只用 2 字节来表示宽度和高度，如果后续有需要，可以考虑改进。

## Webp

开头: 12 字节标识 webp 格式

![20190920150556.png](https://raw.githubusercontent.com/kikt-blog/image/master/img/20190920150556.png)

RIFF(4 字节) + 文件长度(4 字节) + WEBP(4 字节), 文件长度不包含 RIFF 标识头和长度信息, 但包含 WEBP

比如我的图片文件大小为 125,198 字节, 这个文件长度对应的十六进制为 06 E9 01 00 , 这里要注意, 和 png jpg 的长度表示方案不同, 这里是倒序的 0x0001e906 = 125190

---

VP8 chunk: 找到 VP8 的标识 根据内容不同, 表示也不尽相同
往后偏移 `+14 ~ +15` 为宽度(倒序), `16~17`为高度(倒序))

## Gif

开头: `0x47, 0x49, 0x46, 0x38, 0x39, 0x61`
结尾: `0x3B`

从偏移量 `+6~+7` 两位标识宽度 `+8~+9`标识高度, 这里也是倒序, 示例中图片的宽度对应的十六进制为: `[0xB0,0x02] (688)`, 高度为`[0x2E,0x05](1326)`


library/CHANGELOG.md
# CHANGELOG

- [CHANGELOG](#changelog)
  - [2.1.1](#211)
  - [2.1.0](#210)
  - [2.0.0](#200)
  - [1.1.0](#110)
  - [1.0.0](#100)
  - [0.3.0](#030)
  - [0.2.1](#021)
  - [0.2.0](#020)
  - [0.1.1](#011)
  - [0.1.0](#010)

## 2.1.1

- fix: remove debug value print calls from `jpeg_decoder` by @ZaLiTHkA in <https://github.com/CaiJingLong/dart_image_size_getter/pull/21>
- @ZaLiTHkA made their first contribution in <https://github.com/CaiJingLong/dart_image_size_getter/pull/21>

**Full Changelog**: <https://github.com/CaiJingLong/dart_image_size_getter/compare/library-2.1.0...library-2.1.1>

## 2.1.0

Feature:

- Add `needRotate` for `Size`

**What's Changed:** - Add needRotate for Size by @CaiJingLong in <https://github.com/CaiJingLong/dart_image_size_getter/pull/20>

**Full Changelog**: <https://github.com/CaiJingLong/dart_image_size_getter/compare/http_input-2.0.0...library-2.1.0>

## 2.0.0

New feature:

- Refactored the code and removed the `AsyncImageSizeGetter` class.
- We can customize the decoder.

## 1.1.0

- Support webp extended format by @nkming2 in <https://github.com/CaiJingLong/dart_image_size_getter/pull/11>
- Http input by @CaiJingLong in <https://github.com/CaiJingLong/dart_image_size_getter/pull/14>

**New Contributors:** - @nkming2 made their first contribution in <https://github.com/CaiJingLong/dart_image_size_getter/pull/11>

**Full Changelog:** <https://github.com/CaiJingLong/dart_image_size_getter/compare/1.0.0...1.1.0>

## 1.0.0

Null safety version.

## 0.3.0

Support get size of async input of image.

## 0.2.1

Update dependency.
Fix readme.

## 0.2.0

Support memory type.

Breaking change:
Fix: type of `ImageSizeGetter`

## 0.1.1

FIx jpeg format.

## 0.1.0

Initial version.


lib/src/image_size_getter.dart
import 'dart:io';
import 'dart:async';
import 'package:collection/collection.dart';
import 'package:image_size_getter/image_size_getter.dart';
import 'package:image_size_getter/src/decoder/gif_decoder.dart';

import 'package:image_size_getter/src/decoder/jpeg_decoder.dart';
import 'package:image_size_getter/src/decoder/png_decoder.dart';
import 'package:image_size_getter/src/decoder/webp_decoder.dart';
import 'package:image_size_getter/src/utils/file_utils.dart';

class ImageSizGetter {
  static bool isJpg(File file) {
    if (file == null || !file.existsSync()) {
      return false;
    }

    const start = [0xFF, 0xD8];
    const end = [0xFF, 0xD9];

    final util = FileUtils(file);
    final length = file.lengthSync();

    final startList = util.getRangeSync(0, 2);
    final endList = util.getRangeSync(length - 2, length);

    const eq = ListEquality();

    return eq.equals(start, startList) && eq.equals(end, endList);
  }

  static bool isPng(File file) {
    final utils = FileUtils(file);
    final length = file.lengthSync();

    final start = utils.getRangeSync(0, 8);
    final end = utils.getRangeSync(length - 12, length);
    const eq = IterableEquality();
    if (eq.equals(start, _PngHeaders.sig) && eq.equals(end, _PngHeaders.iend)) {
      return true;
    }

    return false;
  }

  static bool isWebp(File file) {
    final utils = FileUtils(file);

    final sizeStart = utils.getRangeSync(0, 4);
    final sizeEnd = utils.getRangeSync(8, 12);

    const eq = ListEquality();

    if (eq.equals(sizeStart, _WebpHeaders.fileSizeStart) &&
        eq.equals(sizeEnd, _WebpHeaders.fileSizeEnd)) {
      return true;
    }
    return false;
  }

  static bool isGif(File file) {
    final utils = FileUtils(file);

    const eq = ListEquality();
    final length = file.lengthSync();

    final sizeStart = utils.getRangeSync(0, 6);
    final sizeEnd = utils.getRangeSync(length - 1, length);

    return eq.equals(sizeStart, _GifHeaders.start) &&
        eq.equals(sizeEnd, _GifHeaders.end);
  }

  static Size getSize(File file) {
    if (isJpg(file)) {
      return JpegDecoder(file).size;
    }
    if (isPng(file)) {
      return PngDecoder(file).size;
    }
    if (isWebp(file)) {
      return WebpDecoder(file).size;
    }
    if (isGif(file)) {
      return GifDecoder(file).size;
    }
    return Size.zero;
  }
}

class _PngHeaders {
  static const sig = [
    0x89,
    0x50,
    0x4E,
    0x47,
    0x0D,
    0x0A,
    0x1A,
    0x0A,
  ];

  static const iend = [
    0x00,
    0x00,
    0x00,
    0x00,
    0x49,
    0x45,
    0x4E,
    0x44,
    0xAE,
    0x42,
    0x60,
    0x82
  ];
}

class _WebpHeaders {
  static const fileSizeStart = [
    0x52,
    0x49,
    0x46,
    0x46,
  ];

  static const fileSizeEnd = [
    0x57,
    0x45,
    0x42,
    0x50,
  ];
}

class _GifHeaders {
  static const start = [
    0x47,
    0x49,
    0x46,
    0x38,
    0x39,
    0x61,
  ];

  static const end = [0x3B];
}
